File: micronaut-app/src/main/java/it/consolemania/games/Game.java
/*
 *   Copyright (c) 2023 Carlo Micieli
 *
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 */
package it.consolemania.games;

import com.jcabi.urn.URN;
import io.micronaut.data.annotation.DateCreated;
import io.micronaut.data.annotation.DateUpdated;
import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.MappedEntity;
import io.micronaut.data.annotation.TypeDef;
import io.micronaut.data.annotation.Version;
import io.micronaut.data.model.DataType;
import io.micronaut.serde.annotation.Serdeable;
import io.micronaut.serde.config.naming.SnakeCaseStrategy;
import it.consolemania.config.URNAttributeConverter;
import java.time.Instant;
import java.util.List;
import java.util.UUID;
import javax.persistence.Embedded;

@MappedEntity("games")
@Serdeable(naming = SnakeCaseStrategy.class)
public record Game(
        @Id UUID gameId,
        @TypeDef(type = DataType.STRING, converter = URNAttributeConverter.class) URN gameUrn,
        UUID platformId,
        String title,
        List<Genre> genres,
        List<Mode> modes,
        String series,
        String developer,
        String publisher,
        @Embedded Release release,
        Integer year,
        @DateCreated Instant createdDate,
        @DateUpdated Instant lastModifiedDate,
        @Version Integer version) {}

File: rust-actix-web/src/games/game.rs
use crate::games::genre::Genre;
use crate::games::metadata::GameMetadata;
use crate::games::mode::Mode;
use crate::games::rating::Rating;
use serde_derive::{Deserialize, Serialize};
use urn::Urn;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, Validate)]
pub struct Game {
    pub game_id: Uuid,
    pub game_urn: Urn,
    pub title: String,
    pub genres: Vec<Genre>,
    pub platform: String,
    pub modes: Vec<Mode>,
    pub series: Option<String>,
    pub developer: Option<String>,
    pub publisher: Option<String>,
    pub rating: Option<Rating>,
    pub plot: Option<String>,
    pub year: i32,
    pub metadata: GameMetadata,
}
